example_scripts/download_vhi_parallel.py

```diff
@@ -0,0 +1,162 @@
+"""
+# NOTE: this file has to be run with ipython
+
+ipython example_scripts/download_vhi_parallel.py
+
+TODO:
+- add the output_dir as an argument
+- figure out how to pass multiple arguments through the pool.map() function
+- so unstable - it has to be run from your root esowc_notes directory (FIX THIS)
+"""
+from ftplib import FTP
+from pathlib import Path
+import multiprocessing
+import ipdb
+
+# to pickle more than multiprocessing can
+# https://stackoverflow.com/a/21345423/9940782
+from pathos.multiprocessing import ProcessingPool as Pool
+
+# ------------------------------------------------------------------------------
+# TODO: NEED to put inside one of the functions
+# OUTPUT_DIR = Path(f'/soge-home/projects/crop_yield/esowc_notes/data/vhi2')
+OUTPUT_DIR = Path('data/vhi_demo').absolute()
+if not OUTPUT_DIR.exists():
+    OUTPUT_DIR.mkdir()
+# ------------------------------------------------------------------------------
+
+def get_ftp_filenames():
+    """  get the filenames of interest """
+    with FTP('ftp.star.nesdis.noaa.gov') as ftp:
+        ftp.login()
+        ftp.cwd('/pub/corp/scsb/wguo/data/Blended_VH_4km/VH/')
+
+        # append the filenames to a list
+        listing = []
+        ftp.retrlines("LIST", listing.append)
+        # extract the filename
+        filepaths = [f.split(' ')[-1] for f in listing]
+        # extract only the filenames of interest
+        vhi_files = [f for f in filepaths if ".VH.nc" in f]
+
+    return vhi_files
+
+
+def download_file_from_ftp(ftp_instance, filename, output_filename):
+    """ by having context manager INSIDE function
+    TODO: is this sensible? or have the FTP as an argument to be supplied ?
+    """
+    print(f"Downloading {output_filename}")
+    with open(output_filename,'wb') as lf:
+        ftp_instance.retrbinary("RETR " + filename, lf.write)
+    if output_filename.exists():
+        print(f"Successful Download! {output_filename}")
+    else:
+        print(f"Error Downloading file: {output_filename}")
+
+    return
+
+
+def batch_ftp_request(filenames, output_dir=OUTPUT_DIR):
```

Comment:
currently working with a batch size of 100, so each `FTP` instance downloads 100 files. No problems as of yet